README.txt
Logitech Unifying tool for Linux

See also the article on <https://lekensteyn.nl/logitech-unifying.html>

Logitech documents

I have learned a bit from the kernel source code hid-logitech-dj, but the
"official" Logitech specification (HID++ 1.0) was much more useful. These
documents can be found on <https://lekensteyn.nl/files/logitech/>.


Debuggers
usbmon.awk - initial debugging tool used for tapping usbmon from debugfs
hidraw.c   - successor of usbmon.awk that can parse packets of usb payload.
read-dev-usbmon.c - Reads data from /dev/usbmonX and show interpreted data in a
  more human-readable way.

Note: as a quick-n-dirty hack, I included hidraw.c at some point into the
read-dev-usbmon program. Otherwise, I had no way to show the difference between
a send or receive packet without adding to the same stdout stream. If I included
it in the stderr pipe, then it would be interleaved with stdout in an
unpredictable manner. This means that hidraw.c is currently unusable, it does
not process data correctly.

Usage of USB debugger:
1. Use `lsusb -d 046d:c52b` to determine the bus number. If the output is "Bus
   001 ..", your usb monitor device is at /dev/usbmon1.
2. sudo chgrp $USER /dev/usbmon1
3. sudo chmod g+r /dev/usbmon1
4. ./read-dev-usbmon /dev/usbmon1
5. Profit!


Pairing tool (ltunify)
ltunify allows you to pair new devices, unpair existing devices or view
information for those devices. In order to build the ltunify binary and install
it to `$HOME/bin/ltunify`:

    make ltunify
    make install-home

If you intend to package ltunify or otherwise install it system-wide with a
udevrule, you can use:

    make ltunify
    make install DESTDIR=$pkgdir bindir=/usr/bin udevrulesdir=/lib/udev/rules.d

Once installed, run `ltunify --help` for available options.

Usage of the pairing tool is pretty straight-forward. Example session:

    $ ./ltunify list
    /dev/hidraw0: Permission denied
    Logitech Unifying Receiver device is not accessible.
    Try running this program as root or enable read/write permissions
    for /dev/hidraw0
    $ sudo chgrp $USER /dev/hidraw0 && sudo chmod g+rw /dev/hidraw0 
    $ ./ltunify list
    Devices count: 1
    Connected devices:
    idx=1   Mouse   M525
    $ ./ltunify info 1
    Device index 1
    Mouse
    Name: M525
    Wireless Product ID: 4013
    Serial number: DAFA335E
    Device was unavailable, version information not available.
    $ ./ltunify unpair 1
    Device 0x01 Mouse successfully unpaired
    $ ./ltunify list
    Devices count: 0
    Connected devices:
    $ ./ltunify pair
    Please turn your wireless device off and on to start pairing.
    Found new device, id=0x01 Mouse
    $ ./ltunify list
    Devices count: 1
    Connected devices:
    idx=1   Mouse   M525

TODO
- organize code in multiple files
- simplify code
- HID++ 2.0 debugging (transparent if possible)

~ Peter Wu <lekensteyn@gmail.com>


NEWS
Version 0.3 - 14 June 2020

* ltunify: fix device version reporting for HID++ 1.0 devices.
* ltunify: ignore DJ reports that resulted in messages such as "Dev conn notif
  is expected to be short, got 0x20 instead".
* ltunify: report some more HID++ 2.0 feature names.
* ltunify: fix failures when another event (such as touchpad motions) occur
  between sending the command and reading the response.
* ltunify: fix crash on unrecognised command-line parameter.
* ltunify: add --version option.
* ltunify: add support for Nano Receiver c534 as used by the MK270 combo.

Version 0.2 - 23 July 2013

* ltunify: display HID++ version and supported HID++ 2.0 features for `info`
  option.
* ltunify: print HID++ version of receiver.
* ltunify: support c52f Nano receiver.
* udev: use uaccess to allow seated users to login without joining a group.
* Add various installation targets and variables to Makefile.
* Updated registers.txt and monitoring tools with battery and keyboard
  information.

Version 0.1 - 25 April 2013

* Initial release.


hidpp20.c
#include <stdio.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdint.h>
#include <string.h>

#define SHORT_MESSAGE           0x10
#define LONG_MESSAGE            0x11

struct hidpp2_message {
	u8 report_id;
	u8 device_index;
	u8 feature_index;
#define HIDPP_SET_FUNC(msg, func) ((msg)->func_swId |= ((func) << 4))
	u8 func_swId;
	u8 params[16]; // 3 or 16 params
} __attribute__((__packed__));

struct feature {
	uint16_t featureId;
#define FEAT_TYPE_MASK          0xe0
#define FEAT_TYPE_OBSOLETE      0x80
#define FEAT_TYPE_SWHIDDEN      0x40
#define FEAT_TYPE_RSVD_INTERNAL 0x20
	u8 featureType;
};

#define SOFTWARE_ID 0x04 /* getRandom() -- guaranteed to be random. */

#define FEATURE_INDEX_IROOT 0x00

#define FID_IFEATURESET 0x0001
static
const char *
get_feature_name(uint16_t featureId) {
	/* With '?' prefix are taken from SetPointP/KEMUI.xml */
	switch (featureId) {
	case 0x0000: return "Root";
	case 0x0001: return "FeatureSet";
	case 0x0002: return "FeatureInfo";
	case 0x0003: return "DeviceFwVersion";
	case 0x0005: return "DeviceName";
	case 0x0006: return "DeviceGroups";
	case 0x00C0: return "Dfucontrol";           /* Firmware Update */
	case 0x1000: return "BatteryStatus";
	case 0x1900: return "?SoundNotif";          /* Sound Notification */
	case 0x1920: return "?AudioControls";       /* Audio Controls */
	case 0x1940: return "?VOIP";                /* Internet Telephony */
	case 0x1960: return "?VideoCalling";
	case 0x1980: return "?Backlighting";
	case 0x1981: return "Backlight";
	case 0x1B00: return "ReprogControls";
	case 0x1B01: return "ReprogControlsV2";
	case 0x1B03: return "ReprogControlsV3";
	case 0x1D4B: return "WirelessDeviceStatus"; /* Wireless Update */
	case 0x2000: return "?MouseFeature";        /* Pointing Device Feature */
	case 0x2001: return "LeftRightSwap";
	case 0x2100: return "VerticalScrolling";
	case 0x2120: return "HiResScrolling";
	case 0x2200: return "MousePointer";
	case 0x2510: return "?ProfileMgmt";         /* Profile Management */
	case 0x4000: return "?KeyboardFeature";
	case 0x40A0: return "FnInversion";
	case 0x40A2: return "NewFnInversion";
	case 0x4100: return "Encryption";
	case 0x4301: return "SolarDashboard";
	case 0x4400: return "?DisplayFeature";
	case 0x4520: return "KeyboardLayout";       /* Inactive key */
	case 0x5500: return "?SliderControls";
	case 0x6000: return "?TouchpadFeature";
	case 0x6010: return "TouchpadFwItems";      /* Basic Touchpad settings */
	case 0x6011: return "TouchpadSwItems";      /* Enhanced Touchpad settings */
	case 0x6012: return "TouchpadWin8FwItems";
	case 0x6020: return "?TouchpadTapSelect";   /* Tap to select */
	case 0x6030: return "?DisablePointerAccel"; /* Disable pointer acceleration */
	case 0x6100: return "TouchpadRawXy";
	case 0x6110: return "TouchmouseRawPoints";
	case 0x6120: return "Touchmouse6120";
	case 0x6300: return "?HandwritingRecog";    /* Handwriting recognition" */
	default: return "unknown";
	}
}

/**
 * Initialize common values of a HID++ 2.0 message: report_id, device_index and
 * software ID. Remaining fields: feature_index, func, params.
 */
#define INIT_HIDPP_SHORT(msg, dev_index) \
	memset((msg), 0, sizeof *(msg)); \
	(msg)->report_id = SHORT_MESSAGE; \
	(msg)->device_index = dev_index; \
	(msg)->func_swId = SOFTWARE_ID

static
bool
do_hidpp2_request(int fd, struct hidpp2_message *msg) {
	struct hidpp_message *hid10_message = (struct hidpp_message *) msg;

        if (!do_write(fd, hid10_message)) {
                return false;
        }

	// use do_read_skippy in case there are interleaving notifications
	// sub id is feature index in HID++ 2.0
	if (!do_read_skippy(fd, hid10_message, 0x11, hid10_message->sub_id)) {
		puts("WTF");
		return false;
	}

	return true;
}

// Returns feature index for featureId
static
u8
get_feature(int fd, u8 device_index, uint16_t featureId) {
	struct hidpp2_message msg;
	INIT_HIDPP_SHORT(&msg, device_index);
	msg.feature_index = FEATURE_INDEX_IROOT;
	HIDPP_SET_FUNC(&msg, 0); // GetFeature(featureId)
	msg.params[0] = featureId >> 8;
	msg.params[1] = featureId & 0xFF;
	if (!do_hidpp2_request(fd, &msg)) {
		return 0;
	}
	return msg.params[0];
}

// Returns number of features or 0 on error.
static
u8
get_feature_count(int fd, u8 device_index, u8 ifeatIndex) {
	struct hidpp2_message msg;
	INIT_HIDPP_SHORT(&msg, device_index);
	// XXX: is this variable? Can't it be hard-coded to 0x01?
	msg.feature_index = ifeatIndex;
	HIDPP_SET_FUNC(&msg, 0); // GetCount()
	if (!do_hidpp2_request(fd, &msg)) {
		fprintf(stderr, "Failed to request features count\n");
		return 0;
	}
	return msg.params[0];
}

// Get featureId and type for a given featureIndex.
static
bool
get_featureId(int fd, u8 device_index, u8 ifeatIndex, u8 featureIndex, struct feature *feat) {
	struct hidpp2_message msg;
	INIT_HIDPP_SHORT(&msg, device_index);
	// XXX: is this variable? Can't it be hard-coded to 0x01?
	msg.feature_index = ifeatIndex;
	HIDPP_SET_FUNC(&msg, 1); // GetFeatureId(featureIndex)
	msg.params[0] = featureIndex;
	if (!do_hidpp2_request(fd, &msg)) {
		return false;
	}
	feat->featureId = (msg.params[0] << 8) | msg.params[1];
	feat->featureType = msg.params[2];
	return true;
}

void
hidpp20_print_features(int fd, u8 device_index) {
	u8 i, count, ifeatIndex;

	ifeatIndex = get_feature(fd, device_index, FID_IFEATURESET);
	if (!ifeatIndex) {
		fprintf(stderr, "Failed to get feature information\n");
		return;
	}
	count = get_feature_count(fd, device_index, ifeatIndex);

	printf("Total number of HID++ 2.0 features: %i\n", count);
	for (i = 0; i <= count; i++) {
		struct feature feat;
		if (get_featureId(fd, device_index, ifeatIndex, i, &feat)) {
			printf(" %2i: [%04X] %c%c%c %s\n", i, feat.featureId,
				feat.featureType & FEAT_TYPE_OBSOLETE ? 'O' : ' ',
				feat.featureType & FEAT_TYPE_SWHIDDEN ? 'H' : ' ',
				feat.featureType & FEAT_TYPE_RSVD_INTERNAL ? 'I' : ' ',
				get_feature_name(feat.featureId));
			if (feat.featureType & ~FEAT_TYPE_MASK) {
				printf("Warning: unrecognized feature flags: %#04x\n",
					feat.featureType & ~FEAT_TYPE_MASK);
			}
		} else {
			fprintf(stderr, "Failed to get feature, is device connected?\n");
		}
	}
	puts("(O = obsolete feature; H = SW hidden feature;\n"
		" I = reserved for internal use)");
}


usbmon.awk
#!/usr/bin/gawk -f
# Formats the output of usbmon, assuming Logitech Unifying Receiver protocol
# Usage: sudo cat /sys/kernel/debug/usb/usbmon/5u | usbmon.awk
# (on older gawk versions, use gawk --re-interval -f usbmon.awk)
#
# Author: Peter Wu <lekensteyn@gmail.com>
# Date: 2013-04-04

BEGIN {
	OFS="   ";
# Taken from Linux source, drivers/hid/hid-logitech-dj.h
# Catgegories are taken from patent description of US8386651
# 0x00 - 0x3F HID reports
types["01"] = "KEYBOARD";
types["02"] = "MOUSE";
types["03"] = "CONSUMER_CONTROL";
types["04"] = "SYSTEM_CONTROL";

types["08"] = "MEDIA_CENTER";

types["0E"] = "LEDS";

# 0x40 - 0x7F enumerator notifications
types["40"] = "NOTIF_DEVICE_UNPAIRED";
types["41"] = "NOTIF_DEVICE_PAIRED";
types["42"] = "NOTIF_CONNECTION_STATUS";
types["4A"] = "NOTIF_RECV_LOCK_CHANGED"; # 0100 0000 = ready for connections, 0010 0000 = new connections disabled
types["4B"] = "?NOTIF_PAIR_ACCEPTED"; # 0100 0000

types["7F"] = "NOTIF_ERROR";

# 0x80 - 0xFF enumerator commands; Register Access
types["80"] = "SET_REG"; # CMD_SWITCH
types["81"] = "GET_REG"; # CMD_GET_PAIRED_DEVICES
types["82"] = "SET_LONG_REG";
types["83"] = "GET_LONG_REG";
types["8F"] = "_ERROR_MSG";
# Align type name
maxlen=0;
for (i in types) {
	if (maxlen < length(types[i])) {
		maxlen = length(types[i]);
	}
}

# error messages for type=8F (ERROR_MSG)
errmsgs["01"] = "SUCCESS";
errmsgs["02"] = "INVALID_SUBID";
errmsgs["03"] = "INVALID_ADDRESS";
errmsgs["04"] = "INVALID_VALUE";
errmsgs["05"] = "CONNECT_FAIL";
errmsgs["06"] = "TOO_MANY_DEVICES";
errmsgs["07"] = "ALREADY_EXISTS";
errmsgs["08"] = "BUSY";
errmsgs["09"] = "UNKNOWN_DEVICE";
errmsgs["0a"] = "RESOURCE_ERROR";
errmsgs["0b"] = "REQUEST_UNAVAILABLE";
errmsgs["0c"] = "INVALID_PARAM_VALUE";
errmsgs["0d"] = "WRONG_PIN_CODE";

regs["00"] = "ENABLED_NOTIFS";
regs["02"] = "CONNECTION_STATE";
regs["b2"] = "DEVICE_PAIRING";
regs["b3"] = "DEVICE_ACTIVITY";
regs["b5"] = "PAIRING_INFO";
} # end of BEGIN
function colorize(col, s) {
	return "\033[" col "m" s "\033[m";
}
# global color
function c(s) {
	return colorize(color, s);
}
function endPoint(ep) {
	if (ep == "0") return "output";
	if (ep == "1") return " input";
	if (ep == "2") return colorize("1;33", "enumIf");
	# if (ep == "3") return "   ???"; # seen in the output of usbmon
	return sprintf("%6s", "ep" ep);
}
function dev(hex) {
	if (hex == "ff") {
		return "RECV";
	}
	if (int(hex) >= 1 && int(hex) <= 6) {
		return "DEV" int(hex)
	}
	return "    ";
}
function typeStr(hex) {
	return sprintf("%-" maxlen "s", types[toupper(hex)]);
}
function payload(type, p) {
	v1 = substr(p, 1, 2);
	if (type == "8f") { # error
		er=substr(p, 5, 2);
		reg=substr(p, 3, 2);
		parms = "SubID=" v1
		parms = parms ", Reg=" c(reg) " " regs[reg];
		parms = parms ", er=" c(er);
		parms = parms " " errmsgs[er];
	} else if (type == "80" || type == "81" || type == "82" || type == "83") {
		parms = "reg=" c(v1) " " regs[v1];
		parms = parms " parms=" c(substr(p, 3));
	} else {
		parms = "parms=" c(p);
	}
	return parms;
}

{
	if (match($0, /.*?:[0-9]+:[0-9]{3,}:([0-9]+) .*? = (..)(..)(..)(..) (.*)/, a)) {
		# length 85 is ok for most, but not when starting logitech program
		if (length($0) > 100) {
			print $0;
			$0 = "";
		}
		printf("%-100s", $0);
		color = "1;32";
		# sending data instead of receiving data
		if ($0 ~ " s ") color = "1;31";

		print	" " endPoint(a[1]),
			"report_id=" c(a[2]),
			"dev_idx=" c(a[3]) " " dev(a[3]),
			"type=" c(a[4]) " " typeStr(a[4]),
			payload(a[4], a[5] a[6]);
	} else {
		 print colorize("1;30", $0);
	}
	fflush();
}


read-dev-usbmon.c
/*
 * Tool for reading usbmon messages and writing non-empty data to stdout.
 * Because of limitations of a single output stream, there is currently a hack
 * that directly includes hidraw.c.
 *
 * Copyright (C) 2013 Peter Wu <lekensteyn@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <sys/ioctl.h>
#include <string.h>
#include <stdint.h>
#include <stdlib.h> /* getenv */
#include <errno.h>
#include <sys/time.h> /* gettimeofday */
#include <time.h> /* localtime */

typedef uint16_t u16;
typedef int32_t s32;
typedef uint64_t u64;
typedef int64_t s64;
#define SETUP_LEN 8

/* taken from Linux, Documentation/usb/usbmon.txt */
struct usbmon_packet {
	u64 id;                 /*  0: URB ID - from submission to callback */
	unsigned char type;     /*  8: Same as text; extensible. */
	unsigned char xfer_type; /*    ISO (0), Intr, Control, Bulk (3) */
	unsigned char epnum;    /*     Endpoint number and transfer direction */
	unsigned char devnum;   /*     Device address */
	u16 busnum;             /* 12: Bus number */
	char flag_setup;        /* 14: Same as text */
	char flag_data;         /* 15: Same as text; Binary zero is OK. */
	s64 ts_sec;             /* 16: gettimeofday */
	s32 ts_usec;            /* 24: gettimeofday */
	int status;             /* 28: */
	unsigned int length;    /* 32: Length of data (submitted or actual) */
	unsigned int len_cap;   /* 36: Delivered length */
	union {                 /* 40: */
		unsigned char setup[SETUP_LEN]; /* Only for Control S-type */
		struct iso_rec {                /* Only for ISO */
			int error_count;
			int numdesc;
		} iso;
	} s;
	int interval;           /* 48: Only for Interrupt and ISO */
	int start_frame;        /* 52: For ISO */
	unsigned int xfer_flags; /* 56: copy of URB's transfer_flags */
	unsigned int ndesc;     /* 60: Actual number of ISO descriptors */
};

struct mon_get_arg {
	struct usbmon_packet *hdr;
	void *data;
	size_t alloc;           /* Length of data (can be zero) */
};

#define MON_IOC_MAGIC		0x92
#define MON_IOCQ_URB_LEN	_IO(MON_IOC_MAGIC, 1)
#define MON_IOCX_GET		_IOW(MON_IOC_MAGIC, 6, struct mon_get_arg)

#define NO_MAIN
// HACK - otherwise there is no easy wat to tell whether a packet is read or
// written from the usbmon
#include "hidraw.c"
#undef NO_MAIN

void print_time(void) {
	struct timeval tval;
	struct tm *tm;

	if (gettimeofday(&tval, NULL)) {
		perror("gettimeofday");
		return;
	}
	tm = localtime(&tval.tv_sec);
	printf("%02d:%02d:%02d.%03ld ",
		tm->tm_hour, tm->tm_min, tm->tm_sec,
		tval.tv_usec / 1000);
}

int main(int argc, char ** argv) {
	unsigned char data[1024];
	struct usbmon_packet hdr;
	struct mon_get_arg event;
	int fd, r;

	if (argc < 2) {
		fprintf(stderr, "Usage: %s /dev/usbmonX\n", argv[0]);
		return 1;
	}

	fd = open(argv[1], O_RDONLY);
	if (fd < 0) {
		perror(argv[1]);
		return 1;
	}

	memset(&hdr, 0, sizeof hdr);
	event.hdr = &hdr; // hopefully it is OK to use stack for this
	event.data = &data;
	event.alloc = sizeof data;

	//r = ioctl(fd, MON_IOCQ_URB_LEN);
	//printf("%i\n", r);
	for (;;) {
		memset(&data, 0xCC, sizeof data); // for debugging purposes
		r = ioctl(fd, MON_IOCX_GET, &event);
		if (r == -1 && errno == EINTR) {
			continue;
		}
		if (r < 0) {
			perror("ioctl");
			break;
		}

		// ignore non-data packets
		if (hdr.len_cap) {
			if (getenv("HEX")) {
				unsigned int i;
				printf("Type=%c\n", hdr.type);
				for (i=0; i<hdr.len_cap; i++) {
					printf("%02X%c", data[i],
						i + 1 == hdr.len_cap ? '\n' : ' ');
				}
			} else if (hdr.len_cap > sizeof (struct report)) {
				fprintf(stderr, "Discarding too large packet of length %u!\n", hdr.len_cap);
			} else {
				struct report *report = (struct report *)&data;
				if (hdr.len_cap < 3) {
					fprintf(stderr, "Short data len: %i\n", hdr.len_cap);
					continue;
				}
#define COLOR(c, cstr) "\033[" c "m" cstr "\033[m"
				print_time();
				if (hdr.type == 'C') {
					printf(COLOR("1;32", "Recv\t"));
				} else if (hdr.type == 'S') {
					printf(COLOR("1;31", "Send\t"));
				} else {
					printf(COLOR("1;35", "Type=%c\t") "\n", hdr.type);
				}
				process_msg(report, hdr.len_cap);
				fflush(NULL);
#if 0
				if (write(STDOUT_FILENO, &data, hdr.len_cap) < 0) {
					perror("write");
					break;
				}
#endif
			}
		}
	}

	close(fd);

	return 0;
}


hidraw.c
/*
 * Displays a more human-readable interpretation of the USB data payload
 * for Logitech Unifying Receiver.
 *
 * Example usage: read-dev-usbmon /dev/usbmon0 | hidraw
 *
 * Copyright (C) 2013 Peter Wu <lekensteyn@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdbool.h>

typedef unsigned char u8;

#define SHORT_MSG	0x10
#define SHORT_MSG_LEN	7
#define DJ_SHORT	0x20
#define DJ_SHORT_LEN	15
#define DJ_LONG		0x21
#define DJ_LONG_LEN	32
#define LONG_MSG	0x11
#define LONG_MSG_LEN	20

struct payload_short {
	u8 address;
	u8 value[3];
};
struct payload_long {
	u8 address;
	u8 str[16];
};
struct report {
	u8 report_id;
	u8 device_index;
	u8 sub_id;
	union {
		struct payload_long l;
		struct payload_short s;
	};
} __attribute__((__packed__));

/* types for HID++ report IDs 0x10 and 0x11 */
static const char * report_types[0x100] = {
	[0x00] = "_HIDPP20", // fake type
	// 0x00 - 0x3F HID reports
	[0x01] = "KEYBOARD",
	[0x02] = "MOUSE",
	[0x03] = "CONSUMER_CONTROL",
	[0x04] = "SYSTEM_CONTROL",

	[0x08] = "MEDIA_CENTER",

	[0x0E] = "LEDS",

	// 0x40 - 0x7F enumerator notifications
	[0x40] = "NOTIF_DEVICE_UNPAIRED",
	[0x41] = "NOTIF_DEVICE_PAIRED",
	[0x42] = "NOTIF_CONNECTION_STATUS",
	[0x4A] = "NOTIF_RECV_LOCK_CHANGED",
	[0x4B] = "?NOTIF_PAIR_ACCEPTED",

	[0x7F] = "NOTIF_ERROR",

	// 0x80 - 0xFF enumerator commands; Register Access
	[0x80] = "SET_REG", // was CMD_SWITCH
	[0x81] = "GET_REG", // was CMD_GET_PAIRED_DEVICES
	[0x82] = "SET_LONG_REG",
	[0x83] = "GET_LONG_REG",
	[0x8F] = "_ERROR_MSG",
	[0xFF] = "_HIDPP20_ERROR_MSG",
};

/* types for DJ report IDS 0x20 and 0x21 */
static const char *dj_report_types[0x100] = {
	/* 0x00 - 0x3F: RF reports */
	[0x01] = "KEYBOARD",
	[0x02] = "MOUSE",
	[0x03] = "CONSUMER_CONTROL",
	[0x04] = "SYSTEM_CONTROL",
	[0x08] = "MEDIA_CENTER",
	[0x0E] = "KEYBOARD_LEDS",

	/* 0x40 - 0x7F: DJ notifications */
	[0x40] = "NOTIF_DEVICE_UNPAIRED",
	[0x41] = "NOTIF_DEVICE_PAIRED",
	[0x42] = "NOTIF_CONNECTION_STATUS",

	[0x7F] = "NOTIF_ERROR",

	/* 0x80 - 0xFF: DJ commands */
	[0x80] = "CMD_SWITCH_N_KEEPALIVE",
	[0x81] = "CMD_GET_PAIRED_DEVICES"
};

static const char * error_messages[0x100] = {
	// error messages for type=8F (ERROR_MSG)
	[0x00] = "SUCCESS",
	[0x01] = "INVALID_SUBID",
	[0x02] = "INVALID_ADDRESS",
	[0x03] = "INVALID_VALUE",
	[0x04] = "CONNECT_FAIL",
	[0x05] = "TOO_MANY_DEVICES",
	[0x06] = "ALREADY_EXISTS",
	[0x07] = "BUSY",
	[0x08] = "UNKNOWN_DEVICE",
	[0x09] = "RESOURCE_ERROR",
	[0x0A] = "REQUEST_UNAVAILABLE",
	[0x0B] = "INVALID_PARAM_VALUE",
	[0x0C] = "WRONG_PIN_CODE",
};

// I don't know the upper bound, perhaps 0x10 is enough
static const char * error_messages_hidpp20[0x100] = {
	[0x00] = "NoError",
	[0x01] = "Unknown",
	[0x02] = "InvalidArgument",
	[0x03] = "OutOfRange",
	[0x04] = "HWError",
	[0x05] = "Logitech internal",
	[0x06] = "INVALID_FEATURE_INDEX",
	[0x07] = "INVALID_FUNCTION_ID",
	[0x08] = "Busy",
	[0x09] = "Unsupported",
};

// everything with a '?' is guessed
static const char * registers[0x100] = {
	[0x00] = "ENABLED_NOTIFS",
	[0x01] = "KBD_HAND_DETECT?",
	[0x02] = "CONNECTION_STATE",
	[0x07] = "BATTERY?",
	[0x09] = "FN_KEY_SWAP?",
	[0x17] = "ILLUMINATION_INFO?",
	[0xb2] = "DEVICE_PAIRING",
	[0xb3] = "DEVICE_ACTIVITY",
	[0xb5] = "PAIRING_INFO",
	[0xf1] = "VERSION_INFO?",
};

bool report_type_is_hidpp(u8 report_id) {
	return report_id == SHORT_MSG || report_id == LONG_MSG;
}
bool report_type_is_dj(u8 report_id) {
	return report_id == DJ_SHORT || report_id == DJ_LONG;
}

const char * report_type_str(u8 report_id, u8 type) {
	const char *str = NULL;
	if (report_type_is_hidpp(report_id))
		str = report_types[type];
	else if (report_type_is_dj(report_id))
		str = dj_report_types[type];
	return str ? str : "";
}
const char * device_type_str(u8 type) {
	switch (type) {
	case 0x01: return "DEV1";
	case 0x02: return "DEV2";
	case 0x03: return "DEV3";
	case 0x04: return "DEV4";
	case 0x05: return "DEV5";
	case 0x06: return "DEV6";
	case 0xFF: return "RECV";
	default: return "";
	}
}
const char *error_str(u8 er) {
	const char * str = error_messages[er];
	return str ? str : "";
}
const char *error_str_hidpp20(u8 er) {
	const char * str = error_messages_hidpp20[er];
	return str ? str : "";
}
const char *register_str(u8 reg) {
	const char * str = registers[reg];
	return str ? str : "";
}

void process_msg_payload(struct report *r, u8 data_len) {
	u8 pos, i;
	u8 * bytes = (u8 *) &r->s;

	pos = 0; // nothing has been processed

	if (report_type_is_hidpp(r->report_id))
	switch (r->sub_id) {
	case 0x00: // assume HID++ 2.0 request/response for feature IRoot
		if (data_len == 4 || data_len == 17) {
			printf("func=%X  ", bytes[0] >> 4);
			printf("swId=%X  ", bytes[0] & 0xF);
			pos = 1;
		}
		break;
	case 0xFF: // assume HID++ 2.0 error
		if (data_len == 17) {
			printf("feat=%X  ", bytes[0]);
			printf("func=%X  ", bytes[1] >> 4);
			printf("swId=%X  ", bytes[1] & 0xF);
			printf("err=%02X %s  ", bytes[2], error_str_hidpp20(bytes[2]));
			pos = 3;
		}
		break;
	case 0x8F: // error
		// TODO: length check
		printf("SubID=%02X %s  ", bytes[0], report_type_str(r->report_id, bytes[0]));
		printf("reg=%02X %s  ", bytes[1], register_str(bytes[1]));
		printf("err=%02X %s  ", bytes[2], error_str(bytes[2]));
		pos = 4; // everything is processed
		break;
	case 0x80:
	case 0x81:
	case 0x82: /* long */
	case 0x83: /* long */
		printf("reg=%02X %s  ", bytes[0], register_str(bytes[0]));
		pos = 1;
		break;
	}

	if (pos < data_len) {
		printf("params=");
		//printf("params(len=%02X)=", data_len);
	}
	for (i = 0; pos < data_len; pos++, i++) {
		printf("%02X ", bytes[pos]);
		if (i % 4 == 3 && pos + 1 < data_len) {
			putchar(' ');
		}
	}
}

void process_msg(struct report *report, ssize_t size) {
	const char * report_type;

	switch (report->report_id) {
	case SHORT_MSG:
		report_type = "short";
		if (size != SHORT_MSG_LEN) {
			fprintf(stderr, "Invalid short msg len %zi\n", size);
			return;
		}
		break;
	case LONG_MSG:
		report_type = "long";
		if (size != LONG_MSG_LEN) {
			fprintf(stderr, "Invalid long msg len %zi\n", size);
			return;
		}
		break;
	case DJ_SHORT:
		report_type = "dj_s";
		if (size != DJ_SHORT_LEN) {
			fprintf(stderr, "Invalid DJ short msg len %zi\n", size);
			return;
		}
		break;
	case DJ_LONG:
		report_type = "dj_l";
		if (size != DJ_LONG_LEN) {
			fprintf(stderr, "Invalid DJ long msg len %zi\n", size);
			return;
		}
		break;
	default:
		report_type = "unkn";
		//fprintf(stderr, "Unknown report ID %02x, len=%zi\n", report->report_id, size);
		if (size < 3) {
			return;
		}
		break;
	}

	printf("report_id=%02X %-5s ", report->report_id, report_type);
	printf("device=%02X %-4s ", report->device_index,
			device_type_str(report->device_index));
	printf("type=%02X %-23s ", report->sub_id,
			report_type_str(report->report_id, report->sub_id));

	if (size > 3) {
		process_msg_payload(report, size - 3);
	}
	putchar('\n');
}

#ifndef NO_MAIN
int main(int argc, char ** argv) {
	int fd = STDIN_FILENO;
	ssize_t r;
	struct report report;

	if (argc >= 2 && (fd = open(argv[1], O_RDONLY)) < 0) {
		perror(argv[1]);
		return 1;
	}

	do {
		memset(&report, 0xCC, sizeof report); // for debugging purposes
		r = read(fd, &report, sizeof report);
		if (r > 0) {
			process_msg(&report, r);
		}
	} while (r >= 0);

	if (r < 0) {
		perror("read");
	}

	close(fd);

	return 0;
}
#endif /* ! NO_MAIN */
